File: kafka/kafka-annotation-processor/src/test/java/ru/tinkoff/kora/kafka/annotation/processor/consumer/KafkaListenerRecordTest.java
package ru.tinkoff.kora.kafka.annotation.processor.consumer;

import org.apache.kafka.common.serialization.Deserializer;
import org.assertj.core.api.InstanceOfAssertFactories;
import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.common.Tag;
import ru.tinkoff.kora.kafka.common.config.KafkaConsumerConfig;
import ru.tinkoff.kora.kafka.common.consumer.containers.handlers.KafkaRecordHandler;
import ru.tinkoff.kora.kafka.common.consumer.telemetry.KafkaConsumerTelemetry;

import static org.assertj.core.api.Assertions.assertThat;

public class KafkaListenerRecordTest extends AbstractKafkaListenerAnnotationProcessorTest {
    @Test
    public void testProcessRecord() {
        var handler = compile("""
            public class KafkaListenerClass {
                @KafkaListener("test.config.path")
                public void process(ConsumerRecord<String, String> event) {
                }
            }
            """)
            .handler(String.class, String.class);

        handler.handle(record("test", "test-value"), i -> i
            .assertRecord(0)
            .hasKey("test")
            .hasValue("test-value"));
    }

    @Test
    public void testProcessRecordWithTag() throws NoSuchMethodException {
        compile("""
            public class KafkaListenerClass {
                @KafkaListener("test.config.path")
                public void process(ConsumerRecord<@Tag(KafkaListenerClass.class) String, @Tag(String.class) String> event) {
                }
            }
            """);

        compileResult.assertSuccess();
        var module = compileResult.loadClass("KafkaListenerClassModule");
        var container = module.getMethod("kafkaListenerClassProcessContainer", KafkaConsumerConfig.class, KafkaRecordHandler.class, Deserializer.class, Deserializer.class, KafkaConsumerTelemetry.class);
        var keyDeserializer = container.getParameters()[2];
        var valueDeserializer = container.getParameters()[3];

        var keyTag = keyDeserializer.getAnnotation(Tag.class);
        var valueTag = valueDeserializer.getAnnotation(Tag.class);

        assertThat(keyTag).isNotNull()
            .extracting(Tag::value, InstanceOfAssertFactories.array(Class[].class))
            .isEqualTo(new Class<?>[]{compileResult.loadClass("KafkaListenerClass")});
        assertThat(valueTag).isNotNull()
            .extracting(Tag::value, InstanceOfAssertFactories.array(Class[].class))
            .isEqualTo(new Class<?>[]{String.class});
    }

    @Test
    public void testProcessRecordAnyKeyType() {
        var handler = compile("""
            public class KafkaListenerClass {
                @KafkaListener("test.config.path")
                public void process(ConsumerRecord<?, String> event) {
                }
            }
            """)
            .handler(byte[].class, String.class);

        handler.handle(record("test".getBytes(), "test-value"), i -> i
            .assertRecord(0)
            .hasKey("test".getBytes())
            .hasValue("test-value")
        );
        handler.handle(errorKey("test-value"), i -> i
            .assertRecord(0)
            .hasKeyError()
            .hasValue("test-value")
        );
        handler.handle(errorValue("test".getBytes()), i -> i
            .assertRecord(0)
            .hasKey("test".getBytes())
            .hasValueError());
    }

    @Test
    public void testProcessRecordAndConsumer() {
        var handler = compile("""
            public class KafkaListenerClass {
                @KafkaListener("test.config.path")
                public void process(Consumer<String, String> consumer, ConsumerRecord<String, String> event) {
                }
            }
            """)
            .handler(String.class, String.class);

        handler.handle(record("test", "test-value"), i -> i
            .assertConsumer(0)
            .assertRecord(1)
            .hasKey("test")
            .hasValue("test-value")
        );
        handler.handle(errorKey("test-value"), i -> i
            .assertConsumer(0)
            .assertRecord(1)
            .hasKeyError()
            .hasValue("test-value")
        );
        handler.handle(errorValue("test"), i -> i
            .assertConsumer(0)
            .assertRecord(1)
            .hasKey("test")
            .hasValueError());
    }

    @Test
    public void testProcessRecordAndKeyParseException() {
        var handler = compile("""
            public class KafkaListenerClass {
                @KafkaListener("test.config.path")
                public void process(ConsumerRecord<String, String> event, RecordKeyDeserializationException exception) {
                }
            }
            """)
            .handler(String.class, String.class);

        handler.handle(record("test", "test-value"), i -> i
            .assertNoException(1)
            .assertRecord(0)
            .hasKey("test")
            .hasValue("test-value")
        );
        handler.handle(errorKey("test-value"), i -> i
            .assertKeyException(1)
            .assertRecord(0)
            .hasKeyError()
            .hasValue("test-value")
        );
        handler.handle(errorValue("test"), i -> i
            .assertNoException(1)
            .assertRecord(0)
            .hasKey("test")
            .hasValueError());
    }

    @Test
    public void testProcessRecordAndValueParseException() {
        var handler = compile("""
            public class KafkaListenerClass {
                @KafkaListener("test.config.path")
                public void process(ConsumerRecord<String, String> event, RecordValueDeserializationException exception) {
                }
            }
            """)
            .handler(String.class, String.class);

        handler.handle(record("test", "test-value"), i -> i
            .assertNoException(1)
            .assertRecord(0)
            .hasKey("test")
            .hasValue("test-value")
        );
        handler.handle(errorKey("test-value"), i -> i
            .assertNoException(1)
            .assertRecord(0)
            .hasKeyError()
            .hasValue("test-value")
        );
        handler.handle(errorValue("test"), i -> i
            .assertValueException(1)
            .assertRecord(0)
            .hasKey("test")
            .hasValueError());
    }

    @Test
    public void testProcessRecordAndParseException() {
        var handler = compile("""
            public class KafkaListenerClass {
                @KafkaListener("test.config.path")
                public void process(ConsumerRecord<String, String> event, Exception exception) {
                }
            }
            """)
            .handler(String.class, String.class);

        handler.handle(record("test", "test-value"), i -> i
            .assertNoException(1)
            .assertRecord(0)
            .hasKey("test")
            .hasValue("test-value")
        );
        handler.handle(errorKey("test-value"), i -> i
            .assertKeyException(1)
            .assertRecord(0)
            .hasKeyError()
            .hasValue("test-value")
        );
        handler.handle(errorValue("test"), i -> i
            .assertValueException(1)
            .assertRecord(0)
            .hasKey("test")
            .hasValueError()
        );
    }

    @Test
    public void testProcessRecordAndParseThrowable() {
        var handler = compile("""
            public class KafkaListenerClass {
                @KafkaListener("test.config.path")
                public void process(ConsumerRecord<String, String> event, Throwable exception) {
                }
            }
            """)
            .handler(String.class, String.class);

        handler.handle(record("test", "test-value"), i -> i
            .assertNoException(1)
            .assertRecord(0)
            .hasKey("test")
            .hasValue("test-value")
        );
        handler.handle(errorKey("test-value"), i -> i
            .assertKeyException(1)
            .assertRecord(0)
            .hasKeyError()
            .hasValue("test-value")
        );
        handler.handle(errorValue("test"), i -> i
            .assertValueException(1)
            .assertRecord(0)
            .hasKey("test")
            .hasValueError()
        );
    }

}

File: kafka/kafka-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kafka/symbol/processor/consumer/KafkaListenerRecordTest.kt
package ru.tinkoff.kora.kafka.symbol.processor.consumer

import org.apache.kafka.common.serialization.Deserializer
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import ru.tinkoff.kora.common.Tag
import ru.tinkoff.kora.kafka.common.config.KafkaConsumerConfig
import ru.tinkoff.kora.kafka.common.consumer.containers.handlers.KafkaRecordHandler
import ru.tinkoff.kora.kafka.common.consumer.telemetry.KafkaConsumerTelemetry

class KafkaListenerRecordTest : AbstractKafkaListenerAnnotationProcessorTest() {
    @Test
    fun testProcessRecord() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                fun process(event: ConsumerRecord<String, String>) {
                }
            }
            
            """.trimIndent())
    }

    @Test
    fun testProcessRecordSuspend() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                suspend fun process(event: ConsumerRecord<String, String>) {
                }
            }
            
            """.trimIndent())
    }

    @Test
    @Disabled("Is not supported by ksp yet")
    fun testProcessRecordWithTags() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                fun process(event: ConsumerRecord<@Tag(KafkaListener::class) String, @Tag(String::class) String>) {
                }
            }
            """.trimIndent())

        val module = compileResult.loadClass("KafkaListenerModule")
        val container = module.getMethod("kafkaListenerProcessContainer", KafkaConsumerConfig::class.java, KafkaRecordHandler::class.java, Deserializer::class.java, Deserializer::class.java, KafkaConsumerTelemetry::class.java)
        val keyDeserializer = container.parameters[2]
        val valueDeserializer = container.parameters[3]

        val keyTag = keyDeserializer.getAnnotation(Tag::class.java)
        val valueTag = valueDeserializer.getAnnotation(Tag::class.java)

        Assertions.assertThat(keyTag).isNotNull()
        Assertions.assertThat(keyTag.value.map { it.java }).isEqualTo(listOf(compileResult.loadClass("KafkaListener")))
        Assertions.assertThat(valueTag).isNotNull()
        Assertions.assertThat(valueTag.value.map { it.java }).isEqualTo(listOf(compileResult.loadClass("KafkaListener")))
    }

    @Test
    fun testProcessRecordAnyKeyType() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                fun process(event: ConsumerRecord<*, String>) {
                }
            }
            
            """.trimIndent())
    }

    @Test
    fun testProcessRecordAndConsumer() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                fun process(consumer: Consumer<String, String>, event: ConsumerRecord<String, String>) {
                }
            }
            
            """.trimIndent())
    }

    @Test
    fun testProcessRecordAndKeyParseException() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                fun process(event: ConsumerRecord<String, String>?, exception: RecordKeyDeserializationException?) {
                }
            }
            
            """.trimIndent())
    }

    @Test
    fun testProcessRecordAndValueParseException() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                fun process(event: ConsumerRecord<String, String>?, exception: RecordValueDeserializationException?) {
                }
            }
            
            """.trimIndent())
    }

    @Test
    fun testProcessRecordAndParseException() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                fun process(event: ConsumerRecord<String, String>?, exception: Exception?) {
                }
            }
            
            """.trimIndent())
    }

    @Test
    fun testProcessRecordAndParseThrowable() {
        compile("""
            class KafkaListenerClass {
                @KafkaListener("test.config.path")
                fun process(event: ConsumerRecord<String, String>?, exception: Throwable?) {
                }
            }
            
            """.trimIndent())
    }
}
